### PR TITLE
[DeviceAsan][NFC] Fix ASan e2e tests, enable CPU or GPU tests

### PR DESCRIPTION
**Pull request by AllanZyne**
fix https://github.com/intel/llvm/issues/15461

- Update lit config to make sure tests will run on cpu or gpu device
- disable GEN12 test

### CODE REVIEW

File: sycl/test-e2e/AddressSanitizer/bad-free/bad-free-host.cpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// REQUIRES: linux
+// REQUIRES: linux, cpu || (gpu && level_zero)
```
[the remark this answers is not in the document]
Comment (bader) [reply sycl/test-e2e/AddressSanitizer/bad-free/bad-free-host.cpp:1]: No. Intel NPU device is exposed as Level Zero device as well.

**Comment by aelovikov-intel** [comment]
I see https://github.com/intel/llvm/blob/1ed687b864cecebe6e2f47dd179f2afe97311464/sycl/test-e2e/ESIMD/lit.local.cfg#L3-L4,

why can't `required_features` be used here? If it simply doesn't work, can you fix that instead?

**Comment by sarnex** [comment]
+1 andrei, anything that's used in `UNSUPPORTED` or `REQUIRES` should work the same in the config files.

**Comment by AllanZyne** [comment]
> I see
> 
> https://github.com/intel/llvm/blob/1ed687b864cecebe6e2f47dd179f2afe97311464/sycl/test-e2e/ESIMD/lit.local.cfg#L3-L4
> 
> ,
> why can't `required_features` be used here? If it simply doesn't work, can you fix that instead?

I don't know how to express cpu or gpu on `lit.local.cfg`. If I append "gpu" and "levelzero" into `required_features`, I don't have chance to write "cpu ||".

**Comment by sarnex** [comment]
@AllanZyne Sorry, can you please tell me the conditional that you want to apply to every single asan test? Those I can try to help you do in the cfg file.

For conditions we only want on some tests but not others, we'll need to do that in the test itself.

**Comment by AllanZyne** [comment]
> @AllanZyne Sorry, can you please tell me the condition that you want to apply to every single asan test that isn't there in the cfg file today? Those I can try to help you do in the cfg file.

Thank you!

> For conditions we only want on some tests but not others, we'll need to do that in the test itself.

I think this is my current case.

**Comment by sarnex** [comment]
ok then seems fine from my pov, ill let other code owners review

**Review by wenju-he — APPROVED** [review]
LGTM

**Comment by AllanZyne** [comment]
Hi @aelovikov-intel, did I address your comment?

**Comment by AllanZyne** [comment]
Hi @intel/llvm-reviewers-runtime, please review

**Review by steffenlarsen — APPROVED** [review]
@AllanZyne - Runtime review isn't needed, so you have all the approvals you need.

That said, before we can merge this, you need to remove sycl/test-e2e/AddressSanitizer/nullpointer/global_nullptr.cpp from sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp. That may require runtime approval, so I'll give you that preemptively.

**Comment by AllanZyne** [comment]
> @AllanZyne - Runtime review isn't needed, so you have all the approvals you need.
> 
> That said, before we can merge this, you need to remove sycl/test-e2e/AddressSanitizer/nullpointer/global_nullptr.cpp from sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp and decrement the expected number of unsupported tests in that file. That may require runtime approval, so I'll give you that preemptively.

Thanks, updated.

**Comment by AllanZyne** [comment]
@intel/llvm-gatekeepers, please merge, thanks!

**Comment by AllanZyne** [comment]
If I remember correctly, I pushed the last commit yesterday.
I think the CI test is hang.

**Comment by steffenlarsen** [comment]
> If I remember correctly, I pushed the last commit yesterday. I think the CI test is hang.

CI failed, so I restarted it recently.

**Comment by sarnex** [comment]
Probably there is some hanging test here, we're not seeing this on other PRs. I cancelled the job here to not hog the CI.

@AllanZyne @intel/dpcpp-sanitizers-review Can someone try to reproduce a Gen12 Linux hang introduced by this PR? Thanks

**Comment by yingcong-wu** [comment]
Hi @sarnex @AllanZyne , We noticed that Device Address Sanitizer run into some weird problem with gen 12 and we suspect the problem is with IGC codegen. Therefore I suggest that we skip running on gen 12 for the moment.

**Comment by AllanZyne** [comment]
> Hi @sarnex @AllanZyne , We noticed that Device Address Sanitizer run into some weird problem with gen 12 and we suspect the problem is with IGC codegen. Therefore I suggest that we skip running on gen 12 for the moment.

Sure, thank you for your investigation!

File: sycl/test-e2e/AddressSanitizer/lit.local.cfg [hunk 2]
```diff
@@ -10,9 +10,9 @@ config.substitutions.append(
...
-config.unsupported_features += ['gpu-intel-gen9', 'gpu-intel-gen11', 'gpu-intel-gen12', 'gpu-intel-pvc']
+# https://github.com/intel/llvm/issues/15461
+# Wait for cpu driver uplifting
+config.unsupported_features += ['arch-x86_64']
 
-# GPU testing requires level_zero
-if 'opencl:gpu' in config.sycl_devices:
-   config.required_features += ['level_zero']
+# https://github.com/intel/llvm/issues/15953
+config.unsupported_features += ['arch-intel_gpu_acm_g12']
```
Comment (steffenlarsen) [inline sycl/test-e2e/AddressSanitizer/lit.local.cfg:18]: That's not gen12. It should be something like
```suggestion
config.unsupported_features += ['gpu-intel-gen12']
```

Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/lit.local.cfg:18]: You can see from https://github.com/intel/llvm/issues/15461 that, @sarnex told me 'gpu-intel-gen12' will disable cpu test as well, and @aelovikov-intel said I should use device specific config like "arch-xxxx".

Comment (steffenlarsen) [reply sycl/test-e2e/AddressSanitizer/lit.local.cfg:18]: I see! Good point. In that case, I believe the following should suffice:
```suggestion
config.unsupported_features += ['arch-intel_gpu_12_0_0']
```

Comment (AllanZyne) [reply sycl/test-e2e/AddressSanitizer/lit.local.cfg]: @steffenlarsen Done

**Comment by sarnex** [comment]
Still seeing timeouts on Linux gen12 it seems.

**Comment by AllanZyne** [comment]
> Still seeing timeouts on Linux gen12 it seems.

I have disable GEN12 test now. 
I tested them on GEN12, and it seems there're some issues in igc.

**Comment by AllanZyne** [comment]
Hi @intel/llvm-gatekeepers, please merge. Thanks.

**Pull request merged**